File: core/cyp/agents/analysts.py
"""分析师团：技术面 / 衍生品 / 情绪 / 链上。并行、失败隔离、缺数据即降级。

每位分析师规则可独立产出结构化信号（无 LLM 也能跑）；缺对应数据维度时
标 degraded=True 且置信度压低，由策略官按权重综合。
"""

from __future__ import annotations

from decimal import Decimal
from typing import Protocol

from cyp.agents.base import AgentContext, Vote, blend
from cyp.contracts import AgentId, AnalystReport, MarketSnapshot, Signal
from cyp.data import indicator_snapshot


class Analyst(Protocol):
    id: AgentId

    async def run(self, snap: MarketSnapshot, ctx: AgentContext) -> AnalystReport: ...


class TechnicalAnalyst:
    id: AgentId = "technical"

    async def run(self, snap: MarketSnapshot, ctx: AgentContext) -> AnalystReport:
        ind = indicator_snapshot(snap.ohlcv)
        if ind["last_close"] is None:
            return AnalystReport(agent="technical", stance="neutral", confidence=0.2,
                                 rationale="无 K 线数据", degraded=True)
        votes: list[Vote] = []
        sf, ss = ind["sma_fast"], ind["sma_slow"]
        if sf is not None and ss is not None:
            sign = 1.0 if sf > ss else -1.0
            votes.append(Vote(sign, 1.0, Signal(name="trend", value=f"sma20{'>' if sign > 0 else '<'}sma50")))
        macd, sig = ind["macd"], ind["macd_signal"]
        if macd is not None and sig is not None:
            s = 1.0 if macd > sig else -1.0
            votes.append(Vote(s, 0.8, Signal(name="macd", value=f"{macd:.2f} vs {sig:.2f}")))
        rsi = ind["rsi"]
        if rsi is not None:
            if rsi > 70:
                votes.append(Vote(-1.0, 0.6, Signal(name="rsi", value=f"{rsi:.1f} 超买")))
            elif rsi < 30:
                votes.append(Vote(1.0, 0.6, Signal(name="rsi", value=f"{rsi:.1f} 超卖")))
            else:
                votes.append(Vote(0.0, 0.3, Signal(name="rsi", value=f"{rsi:.1f} 中性")))
        lc, bu, bl = ind["last_close"], ind["bb_upper"], ind["bb_lower"]
        if bu is not None and bl is not None:
            if lc > bu:
                votes.append(Vote(-1.0, 0.4, Signal(name="bollinger", value="上轨外，超伸")))
            elif lc < bl:
                votes.append(Vote(1.0, 0.4, Signal(name="bollinger", value="下轨外，超跌")))
        stance, conf = blend(votes)
        return AnalystReport(agent="technical", stance=stance, confidence=conf,
                             signals=[v.signal for v in votes],
                             rationale=f"技术面：RSI={rsi:.1f} MACD={'金叉' if (macd or 0) > (sig or 0) else '死叉'}"
                             if rsi is not None else "技术面综合")


class DerivativesAnalyst:
    id: AgentId = "derivatives"

    async def run(self, snap: MarketSnapshot, ctx: AgentContext) -> AnalystReport:
        d = snap.derivatives
        if d is None or d.funding_rate is None:
            return AnalystReport(agent="derivatives", stance="neutral", confidence=0.2,
                                 rationale="无衍生品数据", degraded=True)
        votes: list[Vote] = []
        fr = d.funding_rate
        # 资金费为正=多头拥挤（逆向偏空）；为负=空头拥挤（逆向偏多）
        if fr > Decimal("0.0003"):
            votes.append(Vote(-1.0, 0.8, Signal(name="funding", value=f"{fr} 偏高，多头拥挤")))
        elif fr < Decimal("-0.0003"):
            votes.append(Vote(1.0, 0.8, Signal(name="funding", value=f"{fr} 偏低，空头拥挤")))
        else:
            votes.append(Vote(0.0, 0.4, Signal(name="funding", value=f"{fr} 正常")))
        if d.long_short_ratio is not None:
            lsr = d.long_short_ratio
            if lsr > Decimal("1.15"):
                votes.append(Vote(-0.5, 0.5, Signal(name="ls_ratio", value=f"{lsr} 多头偏拥挤")))
            elif lsr < Decimal("0.87"):
                votes.append(Vote(0.5, 0.5, Signal(name="ls_ratio", value=f"{lsr} 空头偏拥挤")))
        stance, conf = blend(votes)
        return AnalystReport(agent="derivatives", stance=stance, confidence=conf,
                             signals=[v.signal for v in votes], rationale=f"衍生品：资金费 {fr}")


class SentimentAnalyst:
    id: AgentId = "sentiment"

    async def run(self, snap: MarketSnapshot, ctx: AgentContext) -> AnalystReport:
        s = snap.sentiment
        if s is None or s.fear_greed is None:
            return AnalystReport(agent="sentiment", stance="neutral", confidence=0.2,
                                 rationale="无情绪数据", degraded=True)
        votes: list[Vote] = []
        fg = s.fear_greed
        # 极端恐惧=逆向偏多；极端贪婪=逆向偏空
        if fg < 25:
            votes.append(Vote(1.0, 0.6, Signal(name="fear_greed", value=f"{fg} 极端恐惧")))
        elif fg > 75:
            votes.append(Vote(-1.0, 0.6, Signal(name="fear_greed", value=f"{fg} 极端贪婪")))
        else:
            votes.append(Vote(0.0, 0.3, Signal(name="fear_greed", value=f"{fg} 中性")))
        if s.news_score is not None:
            votes.append(Vote(float(s.news_score), 0.4, Signal(name="news", value=f"{s.news_score}")))
        stance, conf = blend(votes)
        return AnalystReport(agent="sentiment", stance=stance, confidence=conf,
                             signals=[v.signal for v in votes], rationale=f"情绪：恐贪 {fg}")


class OnchainAnalyst:
    id: AgentId = "onchain"

    async def run(self, snap: MarketSnapshot, ctx: AgentContext) -> AnalystReport:
        o = snap.onchain
        if o is None or o.smart_money_flow is None:
            return AnalystReport(agent="onchain", stance="neutral", confidence=0.2,
                                 rationale="无链上数据（M0 未接入）", degraded=True)
        votes: list[Vote] = []
        smf = o.smart_money_flow
        if smf > 0:
            votes.append(Vote(1.0, 0.8, Signal(name="smart_money", value=f"净流入 {smf}")))
        elif smf < 0:
            votes.append(Vote(-1.0, 0.8, Signal(name="smart_money", value=f"净流出 {smf}")))
        if o.exchange_netflow is not None:
            # 流入交易所=潜在抛压（偏空）
            sign = -1.0 if o.exchange_netflow > 0 else 1.0
            votes.append(Vote(sign, 0.5, Signal(name="exch_netflow", value=f"{o.exchange_netflow}")))
        stance, conf = blend(votes)
        return AnalystReport(agent="onchain", stance=stance, confidence=conf,
                             signals=[v.signal for v in votes], rationale="链上：聪明钱流向")


ANALYSTS: tuple[Analyst, ...] = (
    TechnicalAnalyst(), DerivativesAnalyst(), SentimentAnalyst(), OnchainAnalyst()
)

File: internal/agents/analysts.go
package agents

import (
	"context"
	"fmt"
	"math"
	"sync"

	"github.com/wangchaozhi/cyp-agent/internal/contracts"
)

type Analyst interface {
	ID() contracts.AgentID
	Run(context.Context, contracts.MarketSnapshot, AgentContext) (contracts.AnalystReport, error)
}

type TechnicalAnalyst struct{}
type DerivativesAnalyst struct{}
type SentimentAnalyst struct{}
type OnchainAnalyst struct{}

func (TechnicalAnalyst) ID() contracts.AgentID   { return contracts.AgentTechnical }
func (DerivativesAnalyst) ID() contracts.AgentID { return contracts.AgentDerivatives }
func (SentimentAnalyst) ID() contracts.AgentID   { return contracts.AgentSentiment }
func (OnchainAnalyst) ID() contracts.AgentID     { return contracts.AgentOnchain }

func AllAnalysts() []Analyst {
	return []Analyst{TechnicalAnalyst{}, DerivativesAnalyst{}, SentimentAnalyst{}, OnchainAnalyst{}}
}

// RunAnalysts executes the fixed analyst panel concurrently while preserving
// deterministic report order. A single analyst failure degrades only that
// dimension; parent context cancellation stops the whole panel.
func RunAnalysts(
	ctx context.Context,
	panel []Analyst,
	snapshot contracts.MarketSnapshot,
	agentContext AgentContext,
) (contracts.List[contracts.AnalystReport], error) {
	if panel == nil {
		panel = AllAnalysts()
	}
	reports := make(contracts.List[contracts.AnalystReport], len(panel))
	var wait sync.WaitGroup
	wait.Add(len(panel))
	for index, analyst := range panel {
		index, analyst := index, analyst
		go func() {
			defer wait.Done()
			report, err := analyst.Run(ctx, snapshot, agentContext)
			if err != nil {
				reports[index] = degradedReport(analyst.ID(), "分析失败，已隔离该维度")
				return
			}
			reports[index] = report
		}()
	}
	wait.Wait()
	if err := contextError(ctx); err != nil {
		return nil, err
	}
	return reports, nil
}

func (TechnicalAnalyst) Run(
	ctx context.Context,
	snapshot contracts.MarketSnapshot,
	_ AgentContext,
) (contracts.AnalystReport, error) {
	if err := contextError(ctx); err != nil {
		return contracts.AnalystReport{}, err
	}
	values := indicatorSnapshot(snapshot.OHLCV)
	if values.lastClose == nil {
		return degradedReport(contracts.AgentTechnical, "无 K 线数据"), nil
	}
	votes := make([]Vote, 0, 4)
	if values.smaFast != nil && values.smaSlow != nil {
		sign := -1.0
		operator := "<"
		if *values.smaFast > *values.smaSlow {
			sign, operator = 1, ">"
		}
		votes = append(votes, Vote{Sign: sign, Weight: 1,
			Signal: contracts.Signal{Name: "trend", Value: "sma20" + operator + "sma50"}})
	}
	if values.macd != nil && values.macdSignal != nil {
		sign := -1.0
		if *values.macd > *values.macdSignal {
			sign = 1
		}
		votes = append(votes, Vote{Sign: sign, Weight: 0.8,
			Signal: contracts.Signal{Name: "macd", Value: fmt.Sprintf("%.2f vs %.2f", *values.macd, *values.macdSignal)}})
	}
	if values.rsi != nil {
		vote := Vote{Weight: 0.3, Signal: contracts.Signal{Name: "rsi", Value: fmt.Sprintf("%.1f 中性", *values.rsi)}}
		if *values.rsi > 70 {
			vote.Sign, vote.Weight, vote.Signal.Value = -1, 0.6, fmt.Sprintf("%.1f 超买", *values.rsi)
		} else if *values.rsi < 30 {
			vote.Sign, vote.Weight, vote.Signal.Value = 1, 0.6, fmt.Sprintf("%.1f 超卖", *values.rsi)
		}
		votes = append(votes, vote)
	}
	if values.bbUpper != nil && values.bbLower != nil {
		if *values.lastClose > *values.bbUpper {
			votes = append(votes, Vote{Sign: -1, Weight: 0.4,
				Signal: contracts.Signal{Name: "bollinger", Value: "上轨外，超伸"}})
		} else if *values.lastClose < *values.bbLower {
			votes = append(votes, Vote{Sign: 1, Weight: 0.4,
				Signal: contracts.Signal{Name: "bollinger", Value: "下轨外，超跌"}})
		}
	}
	stance, confidence := Blend(votes)
	rationale := "技术面综合"
	if values.rsi != nil {
		cross := "死叉"
		macd, signal := 0.0, 0.0
		if values.macd != nil {
			macd = *values.macd
		}
		if values.macdSignal != nil {
			signal = *values.macdSignal
		}
		if macd > signal {
			cross = "金叉"
		}
		rationale = fmt.Sprintf("技术面：RSI=%.1f MACD=%s", *values.rsi, cross)
	}
	return reportFromVotes(contracts.AgentTechnical, stance, confidence, votes, rationale), nil
}

func (DerivativesAnalyst) Run(
	ctx context.Context,
	snapshot contracts.MarketSnapshot,
	_ AgentContext,
) (contracts.AnalystReport, error) {
	if err := contextError(ctx); err != nil {
		return contracts.AnalystReport{}, err
	}
	data := snapshot.Derivatives
	if data == nil || data.FundingRate == nil {
		return degradedReport(contracts.AgentDerivatives, "无衍生品数据"), nil
	}
	funding := *data.FundingRate
	votes := make([]Vote, 0, 2)
	switch {
	case funding.Cmp(contracts.MustDecimal("0.0003")) > 0:
		votes = append(votes, Vote{Sign: -1, Weight: 0.8,
			Signal: contracts.Signal{Name: "funding", Value: funding.String() + " 偏高，多头拥挤"}})
	case funding.Cmp(contracts.MustDecimal("-0.0003")) < 0:
		votes = append(votes, Vote{Sign: 1, Weight: 0.8,
			Signal: contracts.Signal{Name: "funding", Value: funding.String() + " 偏低，空头拥挤"}})
	default:
		votes = append(votes, Vote{Weight: 0.4,
			Signal: contracts.Signal{Name: "funding", Value: funding.String() + " 正常"}})
	}
	if data.LongShortRatio != nil {
		ratio := *data.LongShortRatio
		if ratio.Cmp(contracts.MustDecimal("1.15")) > 0 {
			votes = append(votes, Vote{Sign: -0.5, Weight: 0.5,
				Signal: contracts.Signal{Name: "ls_ratio", Value: ratio.String() + " 多头偏拥挤"}})
		} else if ratio.Cmp(contracts.MustDecimal("0.87")) < 0 {
			votes = append(votes, Vote{Sign: 0.5, Weight: 0.5,
				Signal: contracts.Signal{Name: "ls_ratio", Value: ratio.String() + " 空头偏拥挤"}})
		}
	}
	stance, confidence := Blend(votes)
	return reportFromVotes(contracts.AgentDerivatives, stance, confidence, votes,
		"衍生品：资金费 "+funding.String()), nil
}

func (SentimentAnalyst) Run(
	ctx context.Context,
	snapshot contracts.MarketSnapshot,
	_ AgentContext,
) (contracts.AnalystReport, error) {
	if err := contextError(ctx); err != nil {
		return contracts.AnalystReport{}, err
	}
	data := snapshot.Sentiment
	if data == nil || data.FearGreed == nil {
		return degradedReport(contracts.AgentSentiment, "无情绪数据"), nil
	}
	fearGreed := *data.FearGreed
	votes := make([]Vote, 0, 2)
	switch {
	case fearGreed < 25:
		votes = append(votes, Vote{Sign: 1, Weight: 0.6,
			Signal: contracts.Signal{Name: "fear_greed", Value: fmt.Sprintf("%d 极端恐惧", fearGreed)}})
	case fearGreed > 75:
		votes = append(votes, Vote{Sign: -1, Weight: 0.6,
			Signal: contracts.Signal{Name: "fear_greed", Value: fmt.Sprintf("%d 极端贪婪", fearGreed)}})
	default:
		votes = append(votes, Vote{Weight: 0.3,
			Signal: contracts.Signal{Name: "fear_greed", Value: fmt.Sprintf("%d 中性", fearGreed)}})
	}
	if data.NewsScore != nil {
		value, err := data.NewsScore.Float64()
		if err == nil && !math.IsNaN(value) && !math.IsInf(value, 0) {
			votes = append(votes, Vote{Sign: math.Max(-1, math.Min(1, value)), Weight: 0.4,
				Signal: contracts.Signal{Name: "news", Value: data.NewsScore.String()}})
		}
	}
	stance, confidence := Blend(votes)
	return reportFromVotes(contracts.AgentSentiment, stance, confidence, votes,
		fmt.Sprintf("情绪：恐贪 %d", fearGreed)), nil
}

func (OnchainAnalyst) Run(
	ctx context.Context,
	snapshot contracts.MarketSnapshot,
	_ AgentContext,
) (contracts.AnalystReport, error) {
	if err := contextError(ctx); err != nil {
		return contracts.AnalystReport{}, err
	}
	data := snapshot.Onchain
	if data == nil || data.SmartMoneyFlow == nil {
		return degradedReport(contracts.AgentOnchain, "无链上数据（M0 未接入）"), nil
	}
	votes := make([]Vote, 0, 2)
	flow := *data.SmartMoneyFlow
	if flow.IsPositive() {
		votes = append(votes, Vote{Sign: 1, Weight: 0.8,
			Signal: contracts.Signal{Name: "smart_money", Value: "净流入 " + flow.String()}})
	} else if flow.IsNegative() {
		votes = append(votes, Vote{Sign: -1, Weight: 0.8,
			Signal: contracts.Signal{Name: "smart_money", Value: "净流出 " + flow.String()}})
	}
	if data.ExchangeNetflow != nil {
		sign := 1.0
		if data.ExchangeNetflow.IsPositive() {
			sign = -1
		}
		votes = append(votes, Vote{Sign: sign, Weight: 0.5,
			Signal: contracts.Signal{Name: "exch_netflow", Value: data.ExchangeNetflow.String()}})
	}
	stance, confidence := Blend(votes)
	return reportFromVotes(contracts.AgentOnchain, stance, confidence, votes, "链上：聪明钱流向"), nil
}

func degradedReport(agent contracts.AgentID, rationale string) contracts.AnalystReport {
	return contracts.AnalystReport{
		Agent: agent, Stance: contracts.StanceNeutral, Confidence: 0.2,
		Signals: contracts.List[contracts.Signal]{}, Rationale: rationale, Degraded: true,
	}
}

func reportFromVotes(
	agent contracts.AgentID,
	stance contracts.Stance,
	confidence float64,
	votes []Vote,
	rationale string,
) contracts.AnalystReport {
	signals := make(contracts.List[contracts.Signal], 0, len(votes))
	for _, vote := range votes {
		signals = append(signals, vote.Signal)
	}
	return contracts.AnalystReport{
		Agent: agent, Stance: stance, Confidence: confidence,
		Signals: signals, Rationale: rationale,
	}
}

func contextError(ctx context.Context) error {
	if ctx == nil {
		return nil
	}
	select {
	case <-ctx.Done():
		return ctx.Err()
	default:
		return nil
	}
}
